File: apps/mcp-server/src/tools/student.ts
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { z } from 'zod';
import { callFakeServer, formatMcpResponse } from '../utils/api.js';
import { QueryYearSemesterSchema } from '../utils/types.js';

export function registerStudentTools(server: McpServer) {
  // --- Query Tools ---
  server.tool(
    "student_get_profile",
    "Lấy thông tin cá nhân của sinh viên.",
    { token: z.string().describe("Bearer token") },
    async ({ token }) => {
      const result = await callFakeServer("student_get_profile", "GET", "/student/profile", token);
      return formatMcpResponse(result);
    }
  );

  server.tool(
    "student_get_schedule",
    "Xem lịch học học kỳ.",
    { token: z.string().describe("Bearer token"), ...QueryYearSemesterSchema },
    async ({ token, year, semester }) => {
      const result = await callFakeServer("student_get_schedule", "GET", "/student/schedule", token, undefined, { year, semester });
      return formatMcpResponse(result);
    }
  );

  server.tool(
    "student_get_exam_schedule",
    "Xem lịch thi học kỳ.",
    { token: z.string().describe("Bearer token"), ...QueryYearSemesterSchema },
    async ({ token, year, semester }) => {
      const result = await callFakeServer("student_get_exam_schedule", "GET", "/student/schedule/exam", token, undefined, { year, semester });
      return formatMcpResponse(result);
    }
  );

  server.tool(
    "student_get_score",
    "Xem bảng điểm.",
    { token: z.string().describe("Bearer token") },
    async ({ token }) => {
      const result = await callFakeServer("student_get_score", "GET", "/student/score", token);
      return formatMcpResponse(result);
    }
  );

  server.tool(
    "student_get_tuition_fee",
    "Tra cứu học phí.",
    { token: z.string().describe("Bearer token") },
    async ({ token }) => {
      const result = await callFakeServer("student_get_tuition_fee", "GET", "/student/lookup/tuitionfee", token);
      return formatMcpResponse(result);
    }
  );

  server.tool(
    "student_get_insurance",
    "Tra cứu thông tin bảo hiểm y tế.",
    { token: z.string().describe("Bearer token") },
    async ({ token }) => {
      const result = await callFakeServer("student_get_insurance", "GET", "/student/insurance", token);
      return formatMcpResponse(result);
    }
  );

  server.tool(
    "student_get_courses",
    "Lấy danh sách các môn học.",
    { token: z.string().describe("Bearer token") },
    async ({ token }) => {
      const result = await callFakeServer("student_get_courses", "GET", "/student/courses", token);
      return formatMcpResponse(result);
    }
  );

  server.tool(
    "student_get_training_points",
    "Tra cứu điểm rèn luyện.",
    { token: z.string().describe("Bearer token") },
    async ({ token }) => {
      const result = await callFakeServer("student_get_training_points", "GET", "/student/training-points", token);
      return formatMcpResponse(result);
    }
  );

  server.tool(
    "student_get_survey_form",
    "Lấy danh sách form khảo sát sinh viên cần làm.",
    { token: z.string().describe("Bearer token") },
    async ({ token }) => {
      const result = await callFakeServer("student_get_survey_form", "GET", "/student/survey-form", token);
      return formatMcpResponse(result);
    }
  );

  // --- Composite Tools ---
  server.tool(
    "student_get_remaining_credits",
    "Tính số tín chỉ còn thiếu để tốt nghiệp.",
    {
      token: z.string().describe("Bearer token"),
      required_credits: z.number().min(0).describe("Tổng số tín chỉ yêu cầu của chương trình đào tạo")
    },
    async ({ token, required_credits }) => {
      const scoreResult = await callFakeServer("student_get_score", "GET", "/student/score", token);
      if (!scoreResult.ok) {
        return formatMcpResponse({
          ok: false,
          tool: "student_get_remaining_credits",
          error: {
            code: "NOT_FOUND",
            message: "Không đủ dữ liệu điểm để tính toán số tín chỉ tích lũy."
          }
        });
      }
      const earned_credits = scoreResult.data?.earned_credits || 0;
      const remaining = Math.max(required_credits - earned_credits, 0);

      return formatMcpResponse({
        ok: true,
        tool: "student_get_remaining_credits",
        data: {
          required_credits,
          earned_credits,
          remaining_credits: remaining
        }
      });
    }
  );

  // --- Mutation Tools ---
  server.tool(
    "student_create_contact",
    "Gửi form liên hệ.",
    {
      token: z.string().optional().describe("Bearer token (nếu có)"),
      name: z.string().min(1),
      email: z.string().email(),
      phone: z.string().optional(),
      subject: z.string().min(1),
      message: z.string().min(1)
    },
    async (args) => {
      const { token, ...body } = args;
      const result = await callFakeServer("student_create_contact", "POST", "/contact", token, body);
      return formatMcpResponse(result);
    }
  );

  server.tool(
    "student_create_transcript_registration",
    "Đăng ký in bảng điểm.",
    {
      token: z.string().describe("Bearer token"),
      copies: z.number().int().min(1).describe("Số lượng bản in"),
      language: z.enum(["VI", "EN"]).default("VI").describe("Ngôn ngữ"),
      delivery_method: z.enum(["PICKUP", "SHIP"]).describe("Phương thức nhận"),
      shipping_address: z.string().optional().describe("Địa chỉ nhận (bắt buộc nếu SHIP)"),
      phone: z.string().optional().describe("Số điện thoại (bắt buộc nếu SHIP)"),
      note: z.string().optional()
    },
    async (args) => {
      const { token, ...body } = args;
      if (body.delivery_method === "SHIP" && (!body.shipping_address || !body.phone)) {
        return formatMcpResponse({
          ok: false, tool: "student_create_transcript_registration",
          error: { code: "BAD_REQUEST", message: "Yêu cầu cung cấp địa chỉ và số điện thoại khi chọn SHIP." }
        });
      }
      const result = await callFakeServer("student_create_transcript_registration", "POST", "/student/transcript-regis", token, body);
      return formatMcpResponse(result);
    }
  );

  server.tool(
    "student_create_tuition_extend",
    "Xin gia hạn nộp học phí.",
    {
      token: z.string().describe("Bearer token"),
      year: z.number().int().min(1900),
      semester: z.number().int().min(1),
      requested_due_date: z.string().regex(/^\d{4}-\d{2}-\d{2}$/, "Định dạng YYYY-MM-DD"),
      reason: z.string().min(1),
      phone: z.string().min(6).optional(),
      attachments: z.array(z.string()).default([])
    },
    async (args) => {
      const { token, ...body } = args;
      const result = await callFakeServer("student_create_tuition_extend", "POST", "/student/tuition-extend", token, body);
      return formatMcpResponse(result);
    }
  );

  server.tool(
    "student_create_monthly_parking",
    "Đăng ký vé xe tháng.",
    {
      token: z.string().describe("Bearer token"),
      vehicle_type: z.enum(["MOTORBIKE", "CAR", "BICYCLE"]),
      plate_number: z.string().min(3),
      months: z.number().int().min(1).max(12),
      start_month: z.string().regex(/^\d{4}-\d{2}$/, "Định dạng YYYY-MM"),
      owner_name: z.string().optional(),
      note: z.string().optional()
    },
    async (args) => {
      const { token, ...body } = args;
      const result = await callFakeServer("student_create_monthly_parking", "POST", "/student/monthly-parking", token, body);
      return formatMcpResponse(result);
    }
  );

  server.tool(
    "student_create_graduate_registration",
    "Đăng ký xét tốt nghiệp.",
    {
      token: z.string().describe("Bearer token"),
      year: z.number().int().min(1900),
      semester: z.number().int().min(1),
      email: z.string().email(),
      phone: z.string().min(6),
      address: z.string().optional(),
      note: z.string().optional()
    },
    async (args) => {
      const { token, ...body } = args;
      const result = await callFakeServer("student_create_graduate_registration", "POST", "/student/graduate", token, body);
      return formatMcpResponse(result);
    }
  );

  server.tool(
    "student_create_graduation_thesis_registration",
    "Đăng ký khóa luận tốt nghiệp.",
    {
      token: z.string().describe("Bearer token"),
      thesis_title: z.string().min(1),
      advisor_name: z.string().min(1),
      advisor_email: z.string().email().optional(),
      team_members: z.array(z.object({
        student_id: z.string().min(1),
        full_name: z.string().optional()
      })).min(1),
      note: z.string().optional()
    },
    async (args) => {
      const { token, ...body } = args;
      const result = await callFakeServer("student_create_graduation_thesis_registration", "POST", "/student/graduation-thesis", token, body);
      return formatMcpResponse(result);
    }
  );
}

File: apps/mcp-server/internal/tools/student.go
package tools

import (
	"context"
	"fmt"

	"github.com/giakiet05/uit-hub/apps/mcp-server/internal/client"
	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
)

// RegisterStudent registers student-domain query and mutation tools.
func RegisterStudent(s *server.MCPServer, c *client.Client) {
	tokenDesc := "Bearer token nhận từ auth_login"

	yearSemesterOpts := func() []mcp.ToolOption {
		return []mcp.ToolOption{
			mcp.WithString("token", mcp.Required(), mcp.Description(tokenDesc)),
			mcp.WithNumber("year", mcp.Required(), mcp.Description("Năm học (VD: 2024)")),
			mcp.WithNumber("semester", mcp.Required(), mcp.Description("Học kỳ (1, 2, hoặc 3)")),
		}
	}

	// ── get_student_profile ─────────────────────────────────────
	s.AddTool(
		mcp.NewTool("get_student_profile",
			mcp.WithDescription("Lấy thông tin cá nhân của sinh viên đang đăng nhập (họ tên, MSSV, email, SĐT, ngành). Đọc dữ liệu."),
			mcp.WithString("token", mcp.Required(), mcp.Description(tokenDesc)),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			token, err := requireString(req, "token")
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			resp, apiErr := c.Get("/student/profile", token, nil)
			return handleResp(resp, apiErr), nil
		},
	)

	// ── get_student_schedule ────────────────────────────────────
	s.AddTool(
		mcp.NewTool("get_student_schedule",
			append([]mcp.ToolOption{
				mcp.WithDescription("Lấy thời khoá biểu (lịch học) của sinh viên theo năm và học kỳ. Đọc dữ liệu."),
			}, yearSemesterOpts()...)...,
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			token, err := requireString(req, "token")
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			year, _ := optionalNumber(req, "year")
			semester, _ := optionalNumber(req, "semester")
			resp, apiErr := c.Get("/student/schedule", token, map[string]string{
				"year":     fmt.Sprintf("%.0f", year),
				"semester": fmt.Sprintf("%.0f", semester),
			})
			return handleResp(resp, apiErr), nil
		},
	)

	// ── get_exam_schedule ───────────────────────────────────────
	s.AddTool(
		mcp.NewTool("get_exam_schedule",
			append([]mcp.ToolOption{
				mcp.WithDescription("Lấy lịch thi của sinh viên theo năm và học kỳ. Đọc dữ liệu."),
			}, yearSemesterOpts()...)...,
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			token, err := requireString(req, "token")
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			year, _ := optionalNumber(req, "year")
			semester, _ := optionalNumber(req, "semester")
			resp, apiErr := c.Get("/student/schedule/exam", token, map[string]string{
				"year":     fmt.Sprintf("%.0f", year),
				"semester": fmt.Sprintf("%.0f", semester),
			})
			return handleResp(resp, apiErr), nil
		},
	)

	// ── get_student_scores ──────────────────────────────────────
	s.AddTool(
		mcp.NewTool("get_student_scores",
			mcp.WithDescription("Lấy toàn bộ bảng điểm (giữa kỳ, cuối kỳ, tổng) của sinh viên. Đọc dữ liệu."),
			mcp.WithString("token", mcp.Required(), mcp.Description(tokenDesc)),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			token, err := requireString(req, "token")
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			resp, apiErr := c.Get("/student/score", token, nil)
			return handleResp(resp, apiErr), nil
		},
	)

	// ── get_tuition_fee ─────────────────────────────────────────
	s.AddTool(
		mcp.NewTool("get_tuition_fee",
			mcp.WithDescription("Tra cứu học phí hiện tại của sinh viên (số tín chỉ, số tiền, trạng thái). Đọc dữ liệu."),
			mcp.WithString("token", mcp.Required(), mcp.Description(tokenDesc)),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			token, err := requireString(req, "token")
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			resp, apiErr := c.Get("/student/lookup/tuitionfee", token, nil)
			return handleResp(resp, apiErr), nil
		},
	)

	// ── get_insurance ───────────────────────────────────────────
	s.AddTool(
		mcp.NewTool("get_insurance",
			mcp.WithDescription("Tra cứu thông tin bảo hiểm y tế của sinh viên. Đọc dữ liệu."),
			mcp.WithString("token", mcp.Required(), mcp.Description(tokenDesc)),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			token, err := requireString(req, "token")
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			resp, apiErr := c.Get("/student/insurance", token, nil)
			return handleResp(resp, apiErr), nil
		},
	)

	// ── get_enrolled_courses ────────────────────────────────────
	s.AddTool(
		mcp.NewTool("get_enrolled_courses",
			mcp.WithDescription("Lấy danh sách các môn học mà sinh viên đã và đang đăng ký. Đọc dữ liệu."),
			mcp.WithString("token", mcp.Required(), mcp.Description(tokenDesc)),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			token, err := requireString(req, "token")
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			resp, apiErr := c.Get("/student/courses", token, nil)
			return handleResp(resp, apiErr), nil
		},
	)

	// ── get_training_points ─────────────────────────────────────
	s.AddTool(
		mcp.NewTool("get_training_points",
			mcp.WithDescription("Tra cứu điểm rèn luyện của sinh viên (điểm, xếp loại). Đọc dữ liệu."),
			mcp.WithString("token", mcp.Required(), mcp.Description(tokenDesc)),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			token, err := requireString(req, "token")
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			resp, apiErr := c.Get("/student/training-points", token, nil)
			return handleResp(resp, apiErr), nil
		},
	)

	// ── get_survey_forms ────────────────────────────────────────
	s.AddTool(
		mcp.NewTool("get_survey_forms",
			mcp.WithDescription("Lấy danh sách phiếu khảo sát sinh viên cần hoàn thành. Đọc dữ liệu."),
			mcp.WithString("token", mcp.Required(), mcp.Description(tokenDesc)),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			token, err := requireString(req, "token")
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			resp, apiErr := c.Get("/student/survey-form", token, nil)
			return handleResp(resp, apiErr), nil
		},
	)

	// ── create_transcript_request ───────────────────────────────
	s.AddTool(
		mcp.NewTool("create_transcript_request",
			mcp.WithDescription("Đăng ký in bảng điểm. Trả về request_id và trạng thái. Ghi dữ liệu."),
			mcp.WithString("token", mcp.Required(), mcp.Description(tokenDesc)),
			mcp.WithNumber("copies", mcp.Required(), mcp.Description("Số bản in (≥1)")),
			mcp.WithString("language", mcp.Required(), mcp.Description("Ngôn ngữ bảng điểm"), mcp.Enum("VI", "EN")),
			mcp.WithString("delivery_method", mcp.Required(), mcp.Description("Phương thức nhận"), mcp.Enum("PICKUP", "SHIP")),
			mcp.WithString("shipping_address", mcp.Description("Địa chỉ giao (bắt buộc nếu SHIP)")),
			mcp.WithString("phone", mcp.Description("SĐT nhận hàng (bắt buộc nếu SHIP)")),
			mcp.WithString("note", mcp.Description("Ghi chú thêm")),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			token, err := requireString(req, "token")
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			body := buildBody(req, "copies", "language", "delivery_method", "shipping_address", "phone", "note")
			resp, apiErr := c.Post("/student/transcript-regis", token, body)
			return handleResp(resp, apiErr), nil
		},
	)

	// ── create_tuition_extension ────────────────────────────────
	s.AddTool(
		mcp.NewTool("create_tuition_extension",
			mcp.WithDescription("Xin gia hạn nộp học phí. Trả về request_id và trạng thái. Ghi dữ liệu."),
			mcp.WithString("token", mcp.Required(), mcp.Description(tokenDesc)),
			mcp.WithNumber("year", mcp.Required(), mcp.Description("Năm học")),
			mcp.WithNumber("semester", mcp.Required(), mcp.Description("Học kỳ")),
			mcp.WithString("requested_due_date", mcp.Required(), mcp.Description("Ngày xin gia hạn (YYYY-MM-DD)")),
			mcp.WithString("reason", mcp.Required(), mcp.Description("Lý do xin gia hạn")),
			mcp.WithString("phone", mcp.Description("SĐT liên hệ")),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			token, err := requireString(req, "token")
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			body := buildBody(req, "year", "semester", "requested_due_date", "reason", "phone")
			resp, apiErr := c.Post("/student/tuition-extend", token, body)
			return handleResp(resp, apiErr), nil
		},
	)

	// ── create_monthly_parking ──────────────────────────────────
	s.AddTool(
		mcp.NewTool("create_monthly_parking",
			mcp.WithDescription("Đăng ký vé xe tháng. Trả về request_id. Ghi dữ liệu."),
			mcp.WithString("token", mcp.Required(), mcp.Description(tokenDesc)),
			mcp.WithString("vehicle_type", mcp.Required(), mcp.Description("Loại xe"), mcp.Enum("MOTORBIKE", "CAR", "BICYCLE")),
			mcp.WithString("plate_number", mcp.Required(), mcp.Description("Biển số xe")),
			mcp.WithNumber("months", mcp.Required(), mcp.Description("Số tháng đăng ký (1-12)")),
			mcp.WithString("start_month", mcp.Required(), mcp.Description("Tháng bắt đầu (YYYY-MM)")),
			mcp.WithString("owner_name", mcp.Description("Tên chủ xe")),
			mcp.WithString("note", mcp.Description("Ghi chú")),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			token, err := requireString(req, "token")
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			body := buildBody(req, "vehicle_type", "plate_number", "months", "start_month", "owner_name", "note")
			resp, apiErr := c.Post("/student/monthly-parking", token, body)
			return handleResp(resp, apiErr), nil
		},
	)

	// ── create_graduate_request ─────────────────────────────────
	s.AddTool(
		mcp.NewTool("create_graduate_request",
			mcp.WithDescription("Đăng ký xét tốt nghiệp. Trả về request_id. Ghi dữ liệu."),
			mcp.WithString("token", mcp.Required(), mcp.Description(tokenDesc)),
			mcp.WithNumber("year", mcp.Required(), mcp.Description("Năm học")),
			mcp.WithNumber("semester", mcp.Required(), mcp.Description("Học kỳ")),
			mcp.WithString("email", mcp.Required(), mcp.Description("Email liên hệ")),
			mcp.WithString("phone", mcp.Required(), mcp.Description("SĐT liên hệ")),
			mcp.WithString("address", mcp.Description("Địa chỉ")),
			mcp.WithString("note", mcp.Description("Ghi chú")),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			token, err := requireString(req, "token")
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			body := buildBody(req, "year", "semester", "email", "phone", "address", "note")
			resp, apiErr := c.Post("/student/graduate", token, body)
			return handleResp(resp, apiErr), nil
		},
	)

	// ── create_graduation_thesis_request ────────────────────────
	s.AddTool(
		mcp.NewTool("create_graduation_thesis_request",
			mcp.WithDescription("Đăng ký khóa luận tốt nghiệp. Trả về request_id. Ghi dữ liệu."),
			mcp.WithString("token", mcp.Required(), mcp.Description(tokenDesc)),
			mcp.WithString("thesis_title", mcp.Required(), mcp.Description("Tên đề tài")),
			mcp.WithString("advisor_name", mcp.Required(), mcp.Description("Tên giảng viên hướng dẫn")),
			mcp.WithString("advisor_email", mcp.Description("Email giảng viên hướng dẫn")),
			mcp.WithString("note", mcp.Description("Ghi chú")),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			token, err := requireString(req, "token")
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			body := buildBody(req, "thesis_title", "advisor_name", "advisor_email", "note", "team_members")
			resp, apiErr := c.Post("/student/graduation-thesis", token, body)
			return handleResp(resp, apiErr), nil
		},
	)

	// ── create_contact ──────────────────────────────────────────
	s.AddTool(
		mcp.NewTool("create_contact",
			mcp.WithDescription("Gửi form liên hệ/phản hồi tới trường. Không cần token. Ghi dữ liệu."),
			mcp.WithString("name", mcp.Required(), mcp.Description("Họ tên người gửi")),
			mcp.WithString("email", mcp.Required(), mcp.Description("Email liên hệ")),
			mcp.WithString("phone", mcp.Description("SĐT")),
			mcp.WithString("subject", mcp.Required(), mcp.Description("Tiêu đề")),
			mcp.WithString("message", mcp.Required(), mcp.Description("Nội dung")),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			body := buildBody(req, "name", "email", "phone", "subject", "message")
			resp, apiErr := c.Post("/contact", "", body)
			return handleResp(resp, apiErr), nil
		},
	)
}
